Add new package: grpp library

mtaillefumier opened the pull request:
This package has no official version and the repository does not show any recent activities. The dependency is still used by cp2k. 

Important changes:
- patched cmake build system to make it compatible with `cmake` and `pkgconfig`
- Added support of grpp and trexio in cp2k

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

Commit:
Add grpp library

Commit:
Fix formating

Commit:
add grpp and trexio dependencies to cp2k

Commit:
fix branch naming convention

RMeli commented on var/spack/repos/builtin/packages/grpp/package.py:
```diff
@@ -0,0 +1,35 @@
+    homepage = "https://github.com/aoleynichenko/libgrpp"
+    git = "https://github.com/aoleynichenko/libgrpp.git"
+
+    # notify when the package is updated.
```
```suggestion
```

RMeli commented on var/spack/repos/builtin/packages/grpp/package.py:
```diff
@@ -0,0 +1,35 @@
+
+    build_system("cmake")
+
+    version("main", branch="main")
```
Do you think it is worth adding a version based on a git commit hash, to match the version in the CP2K toolchain?

https://github.com/cp2k/cp2k/blob/fa49f9b3c113ba846793c038dd2db98734c87ed3/tools/toolchain/scripts/stage3/install_libgrpp.sh#L9-L11

`main` is a moving target, even though the project is clearly un-maintained at this point.

mtaillefumier replied:
good point. I will add the cp2k version.

mtaillefumier replied:
corrected

RMeli commented on var/spack/repos/builtin/packages/grpp/package.py:
```diff
@@ -0,0 +1,35 @@
+    version("main", branch="main")
+
+    depends_on("c", type="build")  # generated
+    depends_on("fortran", type="build")  # generated
```
```suggestion
    depends_on("c", type="build")
    depends_on("fortran", type="build")
```

mtaillefumier committed:
Update var/spack/repos/builtin/packages/grpp/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/grpp/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

Commit:
Add targeted version

mtaillefumier commented:
@spackbot fix style

mtaillefumier committed:
[@spackbot] updating style on behalf of mtaillefumier

mtaillefumier commented:
unless something is blocking it, would it be possible to merge this PR.

RMeli commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -162,6 +163,11 @@ class Cp2k(MakefilePackage, CMakePackage, CudaPackage, ROCmPackage):
+    depends_on("grpp@2023.12.25", when="+grpp")
+    depends_on("trexio", when="+trexio")
+    # trexio does not depend on grpp but cp2k support of trexio requires it
+    depends_on("grpp", when="+trexio")
```
```suggestion
    depends_on("grpp@2023.12.25", when="+trexio")
```

RMeli commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -996,6 +1002,9 @@ def cmake_args(self):
             self.define_from_variant("CP2K_USE_QUIP", "quip"),
             self.define_from_variant("CP2K_USE_MPI_F08", "mpi_f08"),
+            self.define_from_variant("CP2K_USE_TREXIO", "trexio"),
+            self.define_from_variant("CP2K_USE_GRPP", "grpp"),
+            self.define_from_variant("CP2K_USE_GRPP", "trexio"),
```
I think this needs better handling, otherwise `+grpp ~trexio` will result in conflicting values for `CP2K_USE_GRPP`. Or am I missing something (like order dependence in CMake arguments or similar)?

mtaillefumier replied:
I modified the logic. GRPP is 
- ON when `+grpp`, `+trexio+grpp`
- OFF when `~grpp` independently of the status of `trexio`

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

Commit:
Added extra conflict to fix ~grpp+trexio

RMeli commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -117,7 +117,8 @@ class Cp2k(MakefilePackage, CMakePackage, CudaPackage, ROCmPackage):
     variant("mpi_f08", default=False, description="Use MPI F08 module")
-
+    variant("grpp", default=False, description="Enable libgrrp support", when="@master:")
+    variant("trexio", default=False, description="Enable Trex-IO support", when="@master:")
```
```suggestion
    variant("grpp", default=False, description="Enable libgrrp support", when="@master")
    variant("trexio", default=False, description="Enable Trex-IO support", when="@master")
```
?

RMeli commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -162,6 +163,15 @@ class Cp2k(MakefilePackage, CMakePackage, CudaPackage, ROCmPackage):
+        depends_on("grpp@2023.12.25")
+
+    # the regtests depend on both libraries to be present otherwise they fail
+    conflicts("trexio", when="~grpp")
```
```suggestion
    conflicts("+trexio", when="~grpp")
```

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

msimberg commented on var/spack/repos/builtin/packages/grpp/package.py:
```diff
@@ -0,0 +1,37 @@
+
+    version("main", branch="main")
+    version(
+        "2023.12.25", commit="64d157f1dc95815096b1fd437a5851abeb3425929cf7b2092bf8262db9c5e33d"
```
Just checking, this version number is "made up" (matches the date of the commit), correct? The date format was chosen because it matches what is used in CP2K?

RMeli replied:
I missed it at the first pass, but does this commit actually exist? Looks like the SHA of the package used in CP2K.

The version is indeed what CP2K calls it, but I'm not sure which commit it corresponds to... https://github.com/cp2k/cp2k/blob/26096efc000316683be093141185a0bd14d2043f/tools/toolchain/scripts/stage3/install_libgrpp.sh#L9-L11

msimberg replied:
Good point. It looks this is the sha256 sum of some tarball that cp2k knows how to download, not a commit hash?

msimberg replied:
Yeah, it's the hash of https://www.cp2k.org/static/downloads/libgrpp-main-20231225.zip. @mtaillefumier would you be able to find out if that archive corresponds to some commit in the repo? Otherwise this'd maybe have to use the cp2k archive (suboptimal)?

msimberg replied:
Updating review to block merging until this is resolved.

mtaillefumier replied:
I changed the hash to the specific commit after @RMeli contacted me privately.

Commit:
wrong sha hash

mtaillefumier commented:
would it be possible to sqush merge all these commits.

RMeli commented on var/spack/repos/builtin/packages/grpp/package.py:
```diff
@@ -0,0 +1,35 @@
+
+    license("BSD-3-Clause")
+
+    build_system("cmake")
```
I think this is redundant because it's a `CMakePackage`.

mtaillefumier replied:
build_system or license ? I guess `build-system`

RMeli commented on var/spack/repos/builtin/packages/grpp/package.py:
```diff
@@ -0,0 +1,35 @@
+
+    depends_on("c", type="build")
+    depends_on("fortran", type="build")
+    depends_on("cmake", type="build")
```
I think this is redundant because it's a `CMakePackage`.

mtaillefumier replied:
you mean the depend_on("cmake"). Removed

RMeli commented:
> would it be possible to sqush merge all these commits.

Yes, everything will be squashed.

RMeli commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -119,7 +119,8 @@ class Cp2k(MakefilePackage, CMakePackage, CudaPackage, ROCmPackage):
     variant("mpi_f08", default=False, description="Use MPI F08 module")
     variant("smeagol", default=False, description="Enable libsmeagol support", when="@2025.2:")
-
+    variant("grpp", default=False, description="Enable libgrpp support", when="@master")
```
```suggestion
    variant("grpp", default=False, description="Enable libgrpp support", when="@2025.2")
```

RMeli commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -119,7 +119,8 @@ class Cp2k(MakefilePackage, CMakePackage, CudaPackage, ROCmPackage):
     variant("smeagol", default=False, description="Enable libsmeagol support", when="@2025.2:")
-
+    variant("grpp", default=False, description="Enable libgrpp support", when="@master")
+    variant("trexio", default=False, description="Enable Trex-IO support", when="+grpp")
```
I'm not sure this is the correct logic. This makes the `trexio` variant only available when the user specifies `+grpp`. But does the user need to know that GRPP is needed for TrexIO? I think it would be better to allow `+trexio` independently of `+grpp` and then use GRPP as follows:
```python
depends_on("grpp", when="+grpp")
depends_on("grpp", when"+trexio")
```
What do you think?

RMeli replied:
In this case the logic to define the CMake variables also need to be updated to something like
```
self.define("CP2K_USE_GRPP", self.spec.satisfies("+grpp") or self.spec.satisfies("+trexio"))
```

mtaillefumier replied:
trexio code in the regtests depend explicitly on grpp which most likely means the two libraries work in concert. I did not check the trexio code in details but i get an error when I run the regtests with trexio and grpp off.

mtaillefumier replied:
we could use 

```
depends_on("grpp", when="+grpp") 
depends_on("grpp", when"+trexio")
```

but then we have to turn grpp variant on or do some trickery in the cmake section. Let take this road.

RMeli commented on var/spack/repos/builtin/packages/grpp/package.py:
```diff
@@ -0,0 +1,32 @@
+
+    maintainers("mtaillefumier")
+
+    license("MIT-License")
```
```suggestion
    license("MIT", checked_by="mtaillefumier")
```
I think the [SPDX license identifier for the MIT license](https://spdx.org/licenses/MIT.html) is just `MIT`.

Commit:
minor changes in the patch changing the cmake project name

mtaillefumier committed:
Update var/spack/repos/builtin/packages/grpp/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

Commit:
Change the dependency logic for grpp and trexio

mtaillefumier commented:
I used this trick to turn on grpp variant when trexio is on
```
variant("grpp", default=False, description="Enable libgrpp support", when="@2025.2:")
variant("trexio", default=False, description="Enable Trex-IO support", when="@2025.2:")
variant("grpp", default=True, description="Enable Trex-IO support", when="@2025.2: +trexio")
```

mtaillefumier commented:
@spackbot fix style

Commit:
grpp and trexio can be supported separately in cp2k

Two cp2k tests need the combination +trexio+grpp to work but
it is not explicitly indicated. It will be fixed in cp2k main repository.

grpp and trexio are independent in practice and are treated as such

msimberg reviewed (approved):
Looks fine to me, but I don't know much about the CP2K dependencies. @mtaillefumier can you confirm that with the last commit there is intentionally no longer any relation between grpp and trexio support?

RMeli commented:
> can you confirm that with the last commit there is intentionally no longer any relation between grpp and trexio support

Yes, the apparent dependency was only due to some tests and have been fixed in https://github.com/cp2k/cp2k/pull/3951

RMeli reviewed (changes requested):
Blocking because the `grpp-cmake.patch` doesn't seem to play well with CP2K yet.

mtaillefumier commented:
I am closing this PR because the code is been integrated directly in CP2K.
